### PR TITLE
[refactor][index] Handle ObjC method names with empty selector pieces

### PR DESCRIPTION
**Pull request by hyp**
1st commit: will go to upstream.
The other two: on GitHub.com/apple/swift-clang.

### CODE REVIEW

File: lib/Tooling/Refactor/RenameIndexedFile.cpp [hunk 1]
```diff
@@ -81,7 +81,9 @@ static MatchKind checkOccurrence(const IndexedOccurrence &Occurrence,
   if (BeginLoc.isInvalid())
     return MatchKind::None;
   StringRef SymbolNameStart = Symbol.Name[0];
-  SourceLocation EndLoc = BeginLoc.getLocWithOffset(SymbolNameStart.size());
+  SourceLocation EndLoc = BeginLoc.getLocWithOffset(std::max(
+      SymbolNameStart.size(),
+      size_t(1))); // Take empty Objective-C selector pieces into account.
   if (!SM.isBeforeInTranslationUnit(BeginLoc, EndLoc)) {
```
Comment (benlangmuir) [inline lib/Tooling/Refactor/RenameIndexedFile.cpp]: Is `isBeforeInTranslationUnit` still correct after the above change?

Comment (hyp) [reply lib/Tooling/Refactor/RenameIndexedFile.cpp]: Yeah, the location just after ':' will be after the location of ':'.

Comment (benlangmuir) [reply lib/Tooling/Refactor/RenameIndexedFile.cpp]: I guess what I was asking is: won't it always be after, since the offset >= 1 now instead of >= 0?

Comment (hyp) [reply lib/Tooling/Refactor/RenameIndexedFile.cpp]: Ah. Right. It looks like we don't need this check at all now. The initial implementation used `Lexer::getSourceText(
      CharSourceRange::getCharRange(BeginLoc, EndLoc)` so we needed to verify the range, but now we just lex the token so this is redundant.

**Comment by hyp** [comment]
Updated.

**Review by benlangmuir — APPROVED** [review]
LGTM

**Comment by hyp** [comment]
First commit - r307901

**Comment by hyp** [comment]
To https://github.com/apple/swift-clang
   2bafda8fb0..13bfdc79c2  upstream-with-swift -> upstream-with-swift

**Pull request closed without merging**